### PR TITLE
Pdf generator2.0

### PR DESCRIPTION
**Pull request by TayKaiJun**
PR Notes:
- Updated openpdf version from 1.3.29 to 1.3.30
- PDFGeneratorUtil now prints multiple numbered pages
    - Each page shows 1 level deeper into each block (i.e. all the sub-blocks of a given block)
    - Added error handling for the exceptions thrown by FileOutputStream
- Created new unit test for PDFGeneratorUtil (checks for creation of an uncorrupted file)

### CODE REVIEW

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 1]
```diff
@@ -1,32 +1,76 @@
...
+      document.setFooter(header)
+
+      def printNode(node: ElkNode): Unit = {
+        val (width, height) = generatePageSize(node)
+        document.setPageSize(new Rectangle(width, height))
+
+        if (document.isOpen) {
+          document.newPage
+        }
+        else {
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:40]:
```suggestion
        } else {
```

General style preference here is this more compact style

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 2]
```diff
@@ -1,32 +1,76 @@
...
+          document.open()
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:41]: This logic seems complex - is there a reason document.open() couldn't be at the top? If so, the reason should be commented.

Is what you're getting at that you don't want to newPage on the first page which creates an empty page? Is there a more direct way to test for this (in the spirit of self-documenting code), can you query if the document is empty? Or add an external flag?

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 3]
```diff
@@ -1,32 +1,76 @@
...
+  private def generatePageSize(node: ElkNode): (Float, Float) = {
+    val width = node.getWidth.toFloat + 2 * ElkNodePainter.margin.toFloat
+    val height = node.getHeight.toFloat + 2 * ElkNodePainter.margin.toFloat
+    (width, height)
+  }
+
+  def generate(content: HierarchyBlock, fileName: String): Unit = {
+    try {
+      val document = new Document()
+      val writer = PdfWriter.getInstance(document, new FileOutputStream(fileName))
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:27]: Not blocking for this PR (so potentially a future, separate clean-up PR), but:
- was it possible to get this to write to an in-memory representation instead of directly to a file, eg pass the OutputStream as an argument to this function?
- and if so, would it make sense to move the try/catch to the most-specific code, perhaps in the calling function that deals with file IO?

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 4]
```diff
@@ -1,32 +1,76 @@
...
+      def printNextHierarchyLevel(block: HierarchyBlock, path: DesignPath = DesignPath()): Unit = {
+        val node = HierarchyGraphElk.HBlockToElkNode(block, path)
+        printNode(node)
 
-    // TODO: make a try..catch for FileOutputStream
-    val writer = PdfWriter.getInstance(document, new FileOutputStream(fileName))
-    document.open()
-    val cb = writer.getDirectContent
-    val graphics = cb.createGraphics(width, height)
+        val nameOrder = ProtoUtil.getNameOrder(block.meta)
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:55]: FYI the latest change (ordered map elements in proto) eliminates this. Now, you can do `block.blocks.asPairs` if you need to iterate, since the proto is now defined as an (ordered) list instead of an unordered map

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 5]
```diff
@@ -1,32 +1,76 @@
...
+        val children: Map[DesignPath, HierarchyBlock] = block.blocks.map {
+          case (name, subblock) => (name, subblock.`type`)
+        }.sortKeysFrom(nameOrder)
+          .collect {
+            case (name, BlockLike.Type.Hierarchy(subblock)) if subblock.blocks.nonEmpty => (path + name, subblock)
+          }.toMap
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:61]: You can chain the foreach directly after this! Since `children` isn't used elsewhere, I don't see a compelling reason to need to separate things out.

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 6]
```diff
@@ -1,32 +1,76 @@
...
 
-    val painter = new ElkNodePainter(rootNode)
-    painter.paintComponent(graphics, Color.white)
+        children.foreach(hBlock => {
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:63]:
```suggestion
        children.foreach { case (path, subblock) => {
```

Using a case statement to unpack the element and give them meaningful names can be a more readable than `._1` and `._2` accessors.

File: src/test/scala/edg_ide/edgir_graph/tests/PDFGeneratorTest.scala [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+          linkPort = Some(ValueExpr.Ref("link", "source"))
+        ))),
+        "connect_sink" -> expr.ValueExpr(expr=expr.ValueExpr.Expr.Connected(expr.ConnectedExpr(
+          blockPort = Some(ValueExpr.Ref("sink", "port")),
+          linkPort = Some(ValueExpr.Ref("link", "sink"))
+        ))),
+      ),
+    )
+
+    PDFGeneratorUtil.generate( blockIr, "unit_test.pdf")
```
Comment (ducky64) [inline src/test/scala/edg_ide/edgir_graph/tests/PDFGeneratorTest.scala:76]:
```suggestion
    PDFGeneratorUtil.generate(blockIr, "unit_test.pdf")
```
minor style nit 🙃

**Review by ducky64 — COMMENTED** [review]
This looks good. A few stylistic suggestions, but the core of this looks great. It still needs to be updated for the latest changes (primarily BerkeleyHCI/PolymorphicBlocks#175 ordered proto), I added an inline comment that should describe what needs to change.

One major change recommendation is to have `PDFGeneratorUtil` return (or generate) an in-memory representation, which I think can separate responsibilities. It doesn't have to be in this PR: I'd actually recommend getting this merged now, then doing a follow-on cleanup PR.

**Comment by TayKaiJun** [comment]
Consideration for next PR:

1. try writing to an in-memory representation instead of directly to a file: eg pass the OutputStream as an argument to this function?
2. If no. 1 works, consider moving the try/catch to the most-specific code, perhaps in the calling function that deals with file IO?
3. Refactor unit test graphs to a HGraphTestUtils?

File: src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala [hunk 8]
```diff
@@ -1,32 +1,76 @@
...
-    val painter = new ElkNodePainter(rootNode)
-    painter.paintComponent(graphics, Color.white)
+        block.blocks.asPairs.map {
+          case (name, subblock) => (name, subblock.`type`)
+        }
+          .collect {
+            case (name, BlockLike.Type.Hierarchy(subblock)) if subblock.blocks.nonEmpty => (path + name, subblock)
+          }.toMap.foreach {
+          case (path, subblock) => printNextHierarchyLevel(subblock, path)
+        }
```
Comment (ducky64) [inline src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala:65]:
```suggestion
        block.blocks.asPairs.map {
          case (name, subblock) => (name, subblock.`type`)
        }.collect {
          case (name, BlockLike.Type.Hierarchy(subblock)) if subblock.blocks.nonEmpty => (path + name, subblock)
        }.toMap.foreach {
          case (path, subblock) => printNextHierarchyLevel(subblock, path)
        }
```

Minor style suggestion, this helps keep similar functionality at similar levels of indentation.
Also, is the `toMap` necessary?

**Review by ducky64 — APPROVED** [review]
Looks good. One last suggestion, and feel free to merge after.
